packages/components/src/scroll-lock/index.js

```diff
@@ -1,114 +1,74 @@
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
+import { useRef, useEffect } from '@wordpress/element';
 
 /**
- * Creates a ScrollLock component bound to the specified document.
- *
- * This function creates a ScrollLock component for the specified document
- * and is exposed so we can create an isolated component for unit testing.
- *
- * @param {Object} args Keyword args.
- * @param {HTMLDocument} args.htmlDocument The document to lock the scroll for.
- * @param {string} args.className The name of the class used to lock scrolling.
- * @return {WPComponent} The bound ScrollLock component.
+ * @param {string} className
+ * @return {(locked: boolean) => void} Function allowing you to set whether scrolling is locked.
  */
-export function createScrollLockComponent( {
-	htmlDocument,
-	className = 'lockscroll',
-} = {} ) {
-	if ( ! htmlDocument && typeof document !== 'undefined' ) {
-		htmlDocument = document;
-	}
-
-	let lockCounter = 0;
-
-	/*
-	 * Setting `overflow: hidden` on html and body elements resets body scroll in iOS.
-	 * Save scroll top so we can restore it after locking scroll.
-	 *
-	 * NOTE: It would be cleaner and possibly safer to find a localized solution such
-	 * as preventing default on certain touchmove events.
-	 */
-	let previousScrollTop = 0;
+function useSetLocked( className ) {
+	/** @type {import('react').MutableRefObject<number>} */
+	const previousScrollTop = useRef( 0 );
```

Comment:
The ref means that each instance of `<ScrollLock />` will have its own `previousScrollTop` instance.

If the scroll is locked by one instance and then unlocked by another, it will fail to restore `previousScrollTop`.

Reply:
1. Instance A mounts. Increments counter from 0 to 1. Calls `setLocked( true )`, stores its local `previousScrollTop`.
2. Instance B mounts. Increments counter from 1 to 2. Does nothing else. `previousScrollTop` is 0.
3. Instance A unmounts. Decrements counter from 2 to 1. Does nothing else. The saved `previousScrollTop` is garbage collected.
4. Instance B unmounts. Decrements counter from 1 to 0. Calls `setLocked( false )`. Reads `previousScrollTop` but it's 0 instead of the correct value stored by instance A.

Reply:
Ah okay, I hadn't thought about the components mounting and unmounting out of order like that, but you're correct there is an issue there. Moving `previousScrollTop` to a module level variable will solve this?

Reply:
> Moving `previousScrollTop` to a module level variable will solve this, right?

Yes, and it's the original behavior before the refactoring :+1:

Reply:
Thanks @jsnajdr!